agent/engine/host_resource_manager.go

```diff
@@ -16,20 +16,311 @@
 package engine
 
 import (
+	"fmt"
+	"sync"
+
 	"github.com/aws/amazon-ecs-agent/agent/ecs_client/model/ecs"
 	"github.com/aws/amazon-ecs-agent/agent/logger"
+	"github.com/aws/amazon-ecs-agent/agent/logger/field"
 	"github.com/aws/amazon-ecs-agent/agent/utils"
 )
 
+// TODO remove this once resource, consume are used
+//lint:file-ignore U1000 Ignore all unused code
+
 // initialHostResource keeps account of each task in
 type HostResourceManager struct {
-	initialHostResource map[string]*ecs.Resource
-	consumedResource    map[string]*ecs.Resource
+	initialHostResource       map[string]*ecs.Resource
+	consumedResource          map[string]*ecs.Resource
+	hostResourceManagerRWLock sync.Mutex
 
 	//task.arn to boolean whether host resources consumed or not
 	taskConsumed map[string]bool
 }
 
+type ResourceNotFoundForTask struct {
+	resource string
+}
+
+func (e *ResourceNotFoundForTask) Error() string {
+	return fmt.Sprintf("no %s in task resources", e.resource)
+}
+
+type ResourceIsNilForTask struct {
+	resource string
+}
+
+func (e *ResourceIsNilForTask) Error() string {
+	return fmt.Sprintf("resource %s is nil in task resources", e.resource)
+}
+
+func (h *HostResourceManager) logResources(msg string, taskArn string) {
+	logger.Debug(msg, logger.Fields{
+		"taskArn":   taskArn,
+		"CPU":       *h.consumedResource["CPU"].IntegerValue,
+		"MEMORY":    *h.consumedResource["MEMORY"].IntegerValue,
+		"PORTS_TCP": h.consumedResource["PORTS_TCP"].StringSetValue,
+		"PORTS_UDP": h.consumedResource["PORTS_UDP"].StringSetValue,
+		"GPU":       *h.consumedResource["GPU"].IntegerValue,
+	})
+}
+
+// Returns if resources consumed or not and error status
+// false, nil -> did not consume, task should stay pending
+// false, err -> resources map has errors, task should fail as cannot schedule with 'wrong' resource map (this basically never happens)
+// true, nil -> successfully consumed, task should progress with task creation
+func (h *HostResourceManager) consume(taskArn string, resources map[string]*ecs.Resource) (bool, error) {
+	h.hostResourceManagerRWLock.Lock()
+	defer h.hostResourceManagerRWLock.Unlock()
+	defer h.logResources("Consumed resources after task consume call", taskArn)
+
+	// Check if already consumed
+	_, ok := h.taskConsumed[taskArn]
+	if ok {
+		// Nothing to do, already consumed, return
+		logger.Info("Resources pre-consumed, continue to task creation", logger.Fields{"taskArn": taskArn})
+		return true, nil
+	}
+
+	ok, err := h.consumable(resources)
```

Comment:
What does it typically mean when we have 1 resource type errored out of the 5 resource types, and the other ones were fine? What will be our action when this happens?

Reply:
This should never happen, but putting in as a check. The task should fail to start with incorrect resource conversion as put in the comment above consume `// false, err -> resources map has errors, task should fail as cannot schedule with 'wrong' resource map (this basically never happens)`